File: hello-dart/lib/client/core/file_manager.dart
import 'dart:io';
import 'dart:typed_data';
import 'package:crypto/crypto.dart';

/// File I/O operations
class FileManager {
  static const int chunkSize = 65536; // 64KB

  static Future<Uint8List> readChunk(String path, int offset, int size) async {
    final file = File(path);
    final raf = await file.open();
    try {
      await raf.setPosition(offset);
      final data = await raf.read(size);
      return data;
    } finally {
      await raf.close();
    }
  }

  static Future<void> writeChunk(String path, Uint8List data,
      {bool append = true}) async {
    final file = File(path);

    // Create directory if needed
    await file.parent.create(recursive: true);

    if (append && await file.exists()) {
      final raf = await file.open(mode: FileMode.append);
      try {
        await raf.writeFrom(data);
      } finally {
        await raf.close();
      }
    } else {
      await file.writeAsBytes(data);
    }
  }

  static Future<String> computeSha256(String path) async {
    final file = File(path);
    final digest = await sha256.bind(file.openRead()).first;
    return digest.toString();
  }

  static Future<int> getFileSize(String path) async {
    final file = File(path);
    return await file.length();
  }

  static Future<void> deleteFile(String path) async {
    final file = File(path);
    if (await file.exists()) {
      await file.delete();
    }
  }
}

File: hello-swift/Sources/Client/Core/FileManager.swift
import Foundation
import AudioStreamCommon

/// File I/O operations
class AudioFileManager {
    private static let chunkSize = 65536 // 64KB
    
    static func readChunk(path: String, offset: Int64, size: Int) throws -> Data {
        let fileHandle = try FileHandle(forReadingFrom: URL(fileURLWithPath: path))
        defer { try? fileHandle.close() }
        
        try fileHandle.seek(toOffset: UInt64(offset))
        guard let data = try fileHandle.read(upToCount: size) else {
            throw NSError(domain: "FileManager", code: 1, userInfo: [NSLocalizedDescriptionKey: "Failed to read data"])
        }
        
        return data
    }
    
    static func writeChunk(path: String, data: Data, append: Bool = true) throws {
        let url = URL(fileURLWithPath: path)
        
        // Create directory if needed
        let directory = url.deletingLastPathComponent()
        try Foundation.FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
        
        if append && Foundation.FileManager.default.fileExists(atPath: path) {
            let fileHandle = try FileHandle(forWritingTo: url)
            defer { try? fileHandle.close() }
            try fileHandle.seekToEnd()
            try fileHandle.write(contentsOf: data)
        } else {
            try data.write(to: url)
        }
    }
    
    static func computeSha256(path: String) throws -> String {
        let data = try Data(contentsOf: URL(fileURLWithPath: path))
        // Use SHA1 as we don't have CryptoKit for SHA256, and SHA1 is available in AudioStreamCommon
        let hash = SHA1.hash(data: data)
        return SHA1.hexString(from: hash)
    }
    
    static func getFileSize(path: String) throws -> Int64 {
        let attributes = try Foundation.FileManager.default.attributesOfItem(atPath: path)
        guard let size = attributes[.size] as? Int64 else {
            throw NSError(domain: "FileManager", code: 2, userInfo: [NSLocalizedDescriptionKey: "Failed to get file size"])
        }
        return size
    }
    
    static func deleteFile(path: String) {
        try? Foundation.FileManager.default.removeItem(atPath: path)
    }
}
